File: Cpp/140.cpp
/*
Given a non-empty string s and a dictionary wordDict containing a list of
non-empty words, add spaces in s to construct a sentence where each word is a
valid dictionary word. Return all such possible sentences.

Note:

The same word in the dictionary may be reused multiple times in the
segmentation. You may assume the dictionary does not contain duplicate words.
Example 1:

Input:
s = "catsanddog"
wordDict = ["cat", "cats", "and", "sand", "dog"]
Output:
[
  "cats and dog",
  "cat sand dog"
]
Example 2:

Input:
s = "pineapplepenapple"
wordDict = ["apple", "pen", "applepen", "pine", "pineapple"]
Output:
[
  "pine apple pen apple",
  "pineapple pen apple",
  "pine applepen apple"
]
Explanation: Note that you are allowed to reuse a dictionary word.
Example 3:

Input:
s = "catsandog"
wordDict = ["cats", "dog", "sand", "and", "cat"]
Output:
[]
*/

class Solution {
private:
    unordered_set<string> words;

public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        words.clear();
        copy(wordDict.begin(), wordDict.end(), inserter(words, words.end()));

        int n = s.size();
        vector<int> dp(n + 1, 0);
        dp[0] = 1;

        for (int i = 1; i <= n; ++i) {
            for (int j = i - 1; j >= 0; --j) {
                if (dp[j]) {
                    if (words.count(s.substr(j, i - j))) {
                        dp[i] = 1;
                        break;
                    }
                }
            }
        }

        vector<string> res;
        if (dp[n]) {
            constructSentence(s, dp, 0, res, "");
        }
        return res;
    }

    void constructSentence(const string& s,
                           const vector<int>& dp,
                           int pos,
                           vector<string>& res,
                           string temp) {
        for (int i = pos + 1; i < dp.size(); ++i) {
            if (dp[i]) {
                string new_word = s.substr(pos, i - pos);
                if (words.count(new_word)) {
                    if (i + 1 == dp.size()) {
                        res.push_back(temp + new_word);
                        return;
                    }
                    constructSentence(s, dp, i, res, temp + new_word + " ");
                }
            }
        }
    }
};

File: Python/140.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        res = []
        n = len(s)
        words = set(wordDict)
        if not n:
            return res
        dp = [False]*(n+1)
        dp[0] = True
        for i in range(1, n+1):
            for j in range(i-1, -1, -1):
                if dp[j] and s[j:i] in words:
                    dp[i] = True
                    break
        if not dp[n]:
            return res
        self.createSentence(s, words, 0, "", res, dp)
        return res
    
    def createSentence(self, s, words, pos, st, res, dp):
        for i in range(pos, len(s)):
            if dp[i+1] and s[pos:i+1] in words:
                if i+1 == len(s):
                    res.append(st+s[pos:])
                else:
                    self.createSentence(s, words, i+1, st+s[pos:i+1]+" ", res, dp)
